File: lexer.py
from enum import StrEnum, auto
from dataclasses import dataclass
from collections import namedtuple
from more_itertools import peekable


class TokenType(StrEnum):
    Ident = auto()
    Keyword = auto()
    Number = auto()
    String = auto()
    Punctuation = auto()
    Error = auto()

    def __repr__(self) -> str:
        return self.__str__().title()


@dataclass
class Token:
    type: TokenType
    value: str
    line: int
    column: int

    def __repr__(self) -> str:
        return f"{self.type}({self.value})"


ErrorToken = Token(TokenType.Error, "", 0, 0)

Keywords = ["struct", "fn", "let"]


class LexerError(Exception):
    pass


def tokenize(code):
    line = 1
    column = 1

    input = peekable(code)
    while input:
        c = next(input)
        if c == "\n":
            line += 1
            column = 1
        elif c.isspace():
            column += 1
        elif c == "-" and input.peek() == ">":
            next(input)
            yield Token(TokenType.Punctuation, "->", line, column)
            column += 2
        elif c == "/" and input.peek() == "/":
            while input.peek() != "\n":
                next(input)
                column += 1
        elif c in [
            "{",
            "}",
            "(",
            ")",
            "[",
            "]",
            ":",
            ";",
            ",",
            ".",
            "=",
            "+",
            "-",
            "*",
            "/",
        ]:
            yield Token(TokenType.Punctuation, c, line, column)
            column += 1
        elif c.isdigit():
            num = c
            while input.peek().isdigit():
                num += next(input)
            if input.peek() == ".":
                num += next(input)
                while input.peek().isdigit():
                    num += next(input)
            yield Token(TokenType.Number, num, line, column)
            column += len(num)
        elif c.isalpha():
            ident = c
            while input.peek().isalnum():
                ident += next(input)
            if ident in Keywords:
                yield Token(TokenType.Keyword, ident, line, column)
            else:
                yield Token(TokenType.Ident, ident, line, column)
            column += len(ident)
        elif c == '"':
            s = ""
            while not (input.peek() == '"' and c != "\\"):
                c = next(input)
                s += c
            next(input)
            yield Token(TokenType.String, s, line, column)
            column += len(s) + 2
        else:
            raise LexerError(f"{c} unexpected on line {line} column {column}")

File: src/lexer.rs
#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Keyword<'a>(pub &'a str);

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Ident<'a>(pub &'a str);

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct NumericLiteral<'a>(pub &'a str);

#[derive(Debug, PartialEq, Eq, Clone, strum::Display)]
pub enum TokenType<'a> {
    Keyword(Keyword<'a>),
    Punctuation(&'a str),
    Ident(Ident<'a>),
    NumericLiteral(NumericLiteral<'a>),
    StringLiteral(&'a str),
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct TokenContext {
    pub line: usize,
    pub column: usize,
}

impl std::fmt::Display for TokenContext {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "line {}, column {}", self.line, self.column)
    }
}

#[derive(Debug, PartialEq, Eq, Clone)]
pub struct Token<'a> {
    pub token: TokenType<'a>,
    pub context: TokenContext,
}

impl std::fmt::Display for Token<'_> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        std::fmt::Display::fmt(&self.token, f)
    }
}

#[derive(Debug, PartialEq, Eq, thiserror::Error)]
#[error("lexer error {description} at {context:?}")]
pub struct LexError {
    pub description: String,
    pub context: TokenContext,
}

// Lexer implementation
pub fn tokenize(input: &str) -> Result<Vec<Token>, LexError> {
    let mut tokens = vec![];
    let mut line = 1;
    let mut column = 1;

    let mut chars = input.chars().peekable();

    let mut i = 0;

    while let Some(ch) = chars.next() {
        match ch {
            ' ' | '\t' => {
                column += 1;
                i += 1
            }
            '\n' => {
                line += 1;
                column = 1;
                i += 1;
            }
            '-' if chars.peek() == Some(&'>') => {
                chars.next();
                tokens.push(Token {
                    token: TokenType::Punctuation(&input[i..i+2]),
                    context: TokenContext { line, column },
                });
                i += 2;
            }
            '(' | ')' | '{' | '}' | ':' | ';' | ',' | '.' | '=' | '+' | '-' | '*' | '/' => {
                tokens.push(Token {
                    token: TokenType::Punctuation(&input[i..i + 1]),
                    context: TokenContext { line, column },
                });
                i += 1;
            }
            '0'..='9' => {
                let mut c = 0;
                while let Some('0'..='9') = chars.peek() {
                    chars.next().unwrap();
                    c += 1;
                }

                // handle decimals
                if let Some('.') = chars.peek() {
                    chars.next().unwrap();
                    c += 1;
                    while let Some('0'..='9') = chars.peek() {
                        chars.next().unwrap();
                        c += 1;
                    }
                }

                tokens.push(Token {
                    token: TokenType::NumericLiteral(NumericLiteral(&input[i..i + c])),
                    context: TokenContext { line, column },
                });
            }
            'a'..='z' | 'A'..='Z' => {
                let mut ident = ch.to_string();
                while let Some('a'..='z' | 'A'..='Z' | '0'..='9' | '_') = chars.peek() {
                    ident.push(chars.next().unwrap());
                }

                // handle any keywords
                let token_type = if ident == "fn" || ident == "struct" {
                    TokenType::Keyword(Keyword(&input[i..i + ident.len()]))
                } else {
                    TokenType::Ident(Ident(&input[i..i + ident.len()]))
                };

                tokens.push(Token {
                    token: token_type,
                    context: TokenContext { line, column },
                });
                i += ident.len();
            }
            '"' => {
                let mut text = String::new();

                let mut last_char = None;
                while let Some(c) = chars.peek() {
                    if *c == '"' && last_char != Some('\\') {
                        chars.next().unwrap();
                        break;
                    }

                    last_char = Some(*c);
                    text.push(chars.next().unwrap());
                }

                tokens.push(Token {
                    token: TokenType::StringLiteral(&input[i..i + text.len()]),
                    context: TokenContext { line, column },
                });
                i += text.len();
            }
            _ => {
                return Err(LexError {
                    description: std::format!("unknown token '{}'", ch),
                    context: TokenContext { line, column },
                })
            }
        }
    }

    Ok(tokens)
}
